### PR TITLE
XY focus

### PR DESCRIPTION
**Pull request by maxkatz6**
## What it is about:

XY navigation is essential for devices with DPad-like controllers, where we don't have classic Tab navigation nor we don't have pointer helping with focusing elements. For example: TVs and consoles, or desktop controlled with a gamepad. It also is often useful with a normal keyboard depending on the layout of elements, where tabbing is not optimal. 

Code is translated and adapted from WinUI 3 source code - https://github.com/microsoft/microsoft-ui-xaml/tree/winui3/release/1.5-stable/dxaml/xcp/components/focus/XYFocus.

API of this PR
Everything is implemented as attached properties:
```c#
public class XYFocus
{
     // property definitions are simplified, all properties are mutable
    public static XYFocusNavigationModes NavigationModes (InputElement obj);
    public static XYFocusNavigationStrategy UpNavigationStrategy (InputElement obj);
    public static XYFocusNavigationStrategy DownNavigationStrategy (InputElement obj);
    public static XYFocusNavigationStrategy LeftNavigationStrategy (InputElement obj);
    public static XYFocusNavigationStrategy RightNavigationStrategy (InputElement obj);

    public static InputElement Up (InputElement obj);
    public static InputElement Down (InputElement obj);
    public static InputElement Left (InputElement obj);
    public static InputElement Right (InputElement obj);
}

public enum XYFocusNavigationModes
{
    Disabled = 0,
    Keyboard = 1,
    Gamepad = 2,
    Remote = 4,
    Enabled = Gamepad | Remote | Keyboard
}
public enum XYFocusNavigationStrategy
{
    Auto,
    Projection = 1,
    NavigationDirectionDistance = 2,
    RectilinearDistance = 3
}
```

See https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.input.xyfocusnavigationstrategy?view=winrt-22621 for visual understanding of the XYFocusNavigationStrategy.

Differences from UWP:
1. UWP has these properties on each Control directly, when we have it as attached properties. (debatable)

Temporary differences from UWP (we planning to implemented)
1. UWP supports Gamepad input handling, we don't yet.
2. Focus engagement is not yet implemented. I need to do more changes in FocusManager code. In this PR I planned to make less invasive changes.
3. UWP provides FocusManager APIs to get next focusable element in a direction, without actually focusing it. As well as it accepts some other options like hint rectangle. In this PR the only focus was to handle keyboard input without refactoring our FocusManager.
4. We do not have ability to focus non-visual elements like TextElement (text block inlines) yet. It will likely be done only in 12.0 as I expect bigger breaking changes along the way.

Possible bug that I am leaving outside of the scope of this PR:
1. RectilinearDistance and NavigationDirectionDistance behavior is a bit different from UWP. But for some reason in actual UWP app I am getting **wrong** navigation results comparing to the official documentation. This PR follows official documentation. And also WinUI source code...I can't debug UWP source code to know the reason of these differences.
2. In some cases Projection navigation gives non-optimal results, possibly because of different manifolds handling.
3. In UWP, you can't leave "scope" of the KeyboardNavigationMode=Enabled tree. Couldn't find out why in time.

Animation

### CODE REVIEW

**Comment by grokys** [comment]
> UWP has these properties on each Control directly, when we have it as attached properties. (debatable)

I think it makes sense to have them as attached properties, personally.

A few issues I've noticed:

- Seems to have broken menu arrow key handling (caught by integration tests)
- When using the "Focus" page in ControlCatalog pressing the left key often results in another page being selected

Other than that the code looks good to me from what I've seen.

File: src/Avalonia.Base/Controls/IInternalScroller.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Runtime.CompilerServices;
+
+namespace Avalonia.Controls.Primitives;
+
+internal interface IInternalScroller
```
Comment (grokys) [inline src/Avalonia.Base/Controls/IInternalScroller.cs:5]: Maybe add a comment here explaining why it's needed with a `TODO12` comment explaining how we may want to refactor this when we can make breaking changes.

File: src/Avalonia.Base/Input/KeyboardNavigationHandler.cs [hunk 2]
```diff
@@ -113,6 +117,20 @@ void OnKeyDown(object? sender, KeyEventArgs e)
                 Move(current, direction, e.KeyModifiers);
                 e.Handled = true;
             }
+            else if (e.Key is Key.Left or Key.Right or Key.Up or Key.Down)
```
Comment (grokys) [inline src/Avalonia.Base/Input/KeyboardNavigationHandler.cs:120]: There are no unit tests covering this block of code: might be worth adding one or two just to exercise this code path?

File: src/Avalonia.Base/Input/Navigation/XYFocusOptions.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Avalonia.Input.Navigation;
+
+internal record XYFocusOptions
```
[the remark this answers is not in the document]
Comment (maxkatz6) [reply src/Avalonia.Base/Input/Navigation/XYFocusOptions.cs:3]: In one of the previous iterations it was used as a dictionary key, not anymore. Can be removed.

File: src/Avalonia.Base/Controls/IInternalScroller.cs (hunk 1 above)
Comment (maxkatz6) [reply src/Avalonia.Base/Controls/IInternalScroller.cs:5]: +

File: src/Avalonia.Base/Input/KeyboardNavigationHandler.cs (hunk 2 above)
Comment (maxkatz6) [reply src/Avalonia.Base/Input/KeyboardNavigationHandler.cs:120]: Yeah, there should be. And this code is a reason why Menu tests failed, haha.

**Comment by maxkatz6** [comment]
> When using the "Focus" page in ControlCatalog pressing the left key often results in another page being selected

Fixed that, so only subtrees with enabled XY navigation are selectable.

Also made a change from UWP by replacing XYFocusKeyboardNavigationMode with XYFocusNavigationModes.
In UWP it was only possible to toggle XY for Keyboard input. When it was always enabled for gamepad. With my changes it's a flags enum now which can enable/disable any key device type.

**Review by grokys — APPROVED** [review]
LGTM and seems to work well!

**Pull request merged**